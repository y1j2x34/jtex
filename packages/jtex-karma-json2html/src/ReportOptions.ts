import { JSONReportData } from '../../../common/jtex-karma-json-report-core';

export interface I18NResource {
    [key: string]: string | I18NResource;
}

export interface ReportOptions {
    lang: string;
    i18nResources: {
        [lang: string]: I18NResource;
    };
    pageTitle: string;
    reportTitle: string;
    forcusOnFailures: boolean;
    data: JSONReportData;
}
