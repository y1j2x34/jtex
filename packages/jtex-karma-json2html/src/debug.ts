import { html } from '.';
import fs from 'fs';
import path from 'path';
import json from './debug/debug.karma-result.json';
import { JSONReportData } from '../../../common/jtex-karma-json-report-core';

const htmlStr = html({
    data: (json as unknown) as JSONReportData,
    collapsed: false,
    jumpToError: true,
    lang: 'zh-CN',
    logoFile: path.resolve(__dirname, '../src/res/logo.svg')
});

fs.writeFileSync(path.resolve(__dirname, '../temp/index.html'), htmlStr);
