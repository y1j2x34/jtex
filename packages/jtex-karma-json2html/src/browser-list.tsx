import * as elements from 'typed-html';
import { translater } from './utils/i18n';
import { ReportOptions } from './ReportOptions';

export function generateBrowserList(options: ReportOptions): string {
    const data = options.data;
    const t = translater(options);
    return (
        <div class="dropdown jtex--dropdown">
            <button class="btn btn-primary dropdown-toggle" data-toggle="dropdown" type="button">
                {t('header.browsers')}
            </button>
            <div class="dropdown-menu switch-browser-dropdown-menu">
                {data.browsers.map(browser => {
                    const browserInfo = browser.browser;
                    return (
                        <a
                            class="dropdown-item switch-browser-menu-item"
                            href="javascript:;"
                            menu-browser={browserInfo.id}
                        >
                            <i class="material-icons text-primary">check</i> {browserInfo.name}
                        </a>
                    );
                })}
            </div>
        </div>
    );
}
