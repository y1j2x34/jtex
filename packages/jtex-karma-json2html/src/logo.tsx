import * as elements from 'typed-html';
import { ReportOptions } from './ReportOptions';
import fs from 'fs';

export function logoHtml(options: ReportOptions) {
    if (!options.logoFile) {
        return '';
    }
    console.info(options.logoFile);
    if (!fs.existsSync(options.logoFile)) {
        return (
            <i class="material-icons align-top" style="width:30px;height:30px;font-size:30px">
                broken_image
            </i>
        );
    }
    const extRegex = /\.(\w+)$/;
    const extExc = extRegex.exec(options.logoFile);
    const ext = extExc ? extExc[1] : '';
    switch (ext) {
        case 'jpeg':
        case 'jpg':
        case 'gif':
        case 'png':
        case 'ico': {
            const base64 = fs.readFileSync(options.logoFile).toString('base64');
            return (
                <img
                    src={'data:image/' + ext + ';base64,' + base64}
                    width="30"
                    height="30"
                    class="d-inline-block align-top"
                />
            );
        }
        case 'svg': {
            const base64 = fs.readFileSync(options.logoFile).toString('base64');
            return (
                <object
                    data={'data:image/svg+xml;base64,' + base64}
                    type="image/svg+xml"
                    style="width:30px;height:30px"
                    class="d-inline-block align-top"
                ></object>
            );
        }
    }
}
