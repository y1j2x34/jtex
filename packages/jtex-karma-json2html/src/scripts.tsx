import * as elements from 'typed-html';
import { ReportOptions } from './ReportOptions';
import code from './res/script.txt';

export function scripts(options: ReportOptions) {
    return [
        <script
            src="https://code.jquery.com/jquery-3.2.1.slim.min.js"
            integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN"
            crossorigin="anonymous"
        ></script>,
        <script
            src="https://unpkg.com/popper.js@1.12.6/dist/umd/popper.js"
            integrity="sha384-fA23ZRQ3G/J53mElWqVJEGJzU0sTs+SvzG8fXVWP+kJQ1lwFAOkcUOysnlKJC33U"
            crossorigin="anonymous"
        ></script>,
        <script
            src="https://unpkg.com/bootstrap-material-design@4.1.1/dist/js/bootstrap-material-design.js"
            integrity="sha384-CauSuKpEqAFajSpkdjv3z9t8E7RlpJ1UP0lKM/+NdtSarroVKu069AlsRPKkFBz9"
            crossorigin="anonymous"
        ></script>,
        <script>{`$(document).ready(function() { $('body').bootstrapMaterialDesign(); });`}</script>,
        <script>var karmaResults = {JSON.stringify(options.data.browsers)};</script>,
        <script>{code}</script>
    ].join('');
}
