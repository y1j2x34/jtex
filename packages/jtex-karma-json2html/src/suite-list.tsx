import * as elements from 'typed-html';
import { SuiteResultInfo, AssertionError } from '../../../common/jtex-karma-json-report-core';
import { ReportOptions } from './ReportOptions';

export function suiteList(options: ReportOptions) {
    return (
        <div class="suite-list" data-spy="scroll">
            {options.data.browsers.map(browserStat => {
                const browserInfo = browserStat.browser;
                const grouped = groupSuitesByName(browserStat.results);
                return (
                    <div data-browser={browserInfo.id}>
                        <div class="suite-card-group">
                            {grouped.map(({ suiteName, cases }) => {
                                return suiteCard(suiteName, cases);
                            })}
                        </div>
                    </div>
                );
            })}
        </div>
    );
}
function groupSuitesByName(suites: SuiteResultInfo[]): Array<{ suiteName: string; cases: SuiteResultInfo[] }> {
    const suiteMap = {};
    suites.forEach(result => {
        const suiteName = result.suite.join('-');
        if (!suiteMap[suiteName]) {
            suiteMap[suiteName] = [];
        }
        suiteMap[suiteName].push(result);
    });
    return Object.keys(suiteMap).map(suiteName => {
        return {
            suiteName: suiteName,
            cases: suiteMap[suiteName]
        };
    });
}

function suiteCard(name: string, cases: SuiteResultInfo[]) {
    return (
        <div class="card">
            <div class="card-body">
                <h5 class="cart-title">{name}</h5>
                {cases.map(suiteCase => {
                    if (suiteCase.success) {
                        return suiteCaseTemplate(suiteCase, {
                            itemClass: 'case-success',
                            alertClass: 'alert-success',
                            icon: 'done'
                        });
                    } else if (suiteCase.skipped) {
                        return suiteCaseTemplate(suiteCase, {
                            itemClass: 'case-skipped',
                            alertClass: 'alert-dark',
                            icon: 'highlight_off'
                        });
                    } else if (suiteCase.pending) {
                        return suiteCaseTemplate(suiteCase, {
                            itemClass: 'case-pending',
                            alertClass: 'alert-secondary',
                            icon: 'hourglass_empty'
                        });
                    } else {
                        return suiteCaseTemplate(suiteCase, {
                            itemClass: 'case-error',
                            alertClass: 'alert-danger',
                            icon: 'error'
                        });
                    }
                })}
            </div>
        </div>
    );
}

function suiteCaseTemplate(suiteCase: SuiteResultInfo, data: TemplateData) {
    return (
        <div class={'suite-case-item ' + data.itemClass}>
            <div class={'alert ' + data.alertClass}>
                <i class="material-icons">{data.icon}</i>
                <span class="suite-case-desc">{suiteCase.description}</span>
                <span class="suite-case-time">
                    <i class="material-icons">update</i>
                    {suiteCase.endTime - suiteCase.startTime}ms
                </span>
            </div>
            {Array.isArray(suiteCase.log) ? logsTemplate(suiteCase.log) : ''}
            {assertionErrorsTemplate(suiteCase.assertionErrors)}
        </div>
    );
}
function logsTemplate(log: string[]) {
    return log
        .map(logItem => {
            return <pre class="suite-case-log">{logItem}</pre>;
        })
        .join('');
}
function assertionErrorsTemplate(assertionErrors: AssertionError[]) {
    return assertionErrors
        .map(assertionError => {
            return (
                <div class="suite-case-assertion-error">
                    <span class="suite-case-assertion-error-message text-danger">{assertionError.message}</span>
                    <span class="suite-case-assertion-error-actual text-danger">{assertionError.actual}</span>
                    <span class="suite-case-assertion-error-expected text-primary">{assertionError.expected}</span>
                </div>
            );
        })
        .join('');
}
interface TemplateData {
    itemClass: string;
    alertClass: string;
    icon: string;
}
