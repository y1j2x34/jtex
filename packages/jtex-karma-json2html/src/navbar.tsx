import * as elements from 'typed-html';
import { generateBrowserList } from './browser-list';
import { translater } from './utils/i18n';
import { ReportOptions } from './ReportOptions';
import { logoHtml } from './logo';
import { searchHtml } from './search';

export function navbar(options: ReportOptions) {
    const t = translater(options);
    return (
        <nav class="navbar navbar-dark bg-primary">
            <a class="navbar-brand" href="javascript:;">
                {logoHtml(options)}
                {t(options.reportTitle)}
            </a>
            <div class="mr-auto">{generateBrowserList(options)}</div>
            <div class="test-summary">
                <span class="badge badge-light">
                    <i class="material-icons">access_alarms</i> <span id="browser-total-time"></span>
                </span>
                <span class="badge badge-light" title={t('header.total-count')}>
                    <i class="material-icons">view_list</i>: <span id="browser-total-count"></span>
                </span>
                <span class="badge badge-success" title={t('header.success-count')}>
                    <i class="material-icons">done</i> <span id="browser-success-count"></span>
                </span>
                <span class="badge badge-danger" title={t('header.failed-count')}>
                    <i class="material-icons">error</i>
                    <span id="browser-error-count"></span>
                </span>
                <span class="badge badge-dark" title={t('header.skipped-count')}>
                    <i class="material-icons">highlight_off</i> <span id="browser-skipped-count"></span>
                </span>
            </div>
        </nav>
    );
}
