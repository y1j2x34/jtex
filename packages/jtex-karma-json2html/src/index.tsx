import * as elements from 'typed-html';
import { JSONReportData } from '../../../common/jtex-karma-json-report-core';
import { generateHeader } from './header';
import { navbar } from './navbar';
import { progressbar } from './progressbar';
import { ReportOptions } from './ReportOptions';
import { scripts } from './scripts';
import { suiteList } from './suite-list';
import { isNotDefined } from './utils/isNotDefined';
import { mergeI18nResources } from './utils/mergeI18nResources';

export function html(opts: Partial<ReportOptions> & { data: JSONReportData }): string {
    if (!opts.lang) {
        opts.lang = 'en-US';
    }
    if (isNotDefined(opts.forcusOnFailures)) {
        opts.forcusOnFailures = false;
    }
    if (isNotDefined(opts.pageTitle)) {
        opts.pageTitle = 'page-title';
    }
    if (isNotDefined(opts.reportTitle)) {
        opts.reportTitle = opts.pageTitle;
    }
    opts.i18nResources = mergeI18nResources(opts.i18nResources);
    const options = opts as ReportOptions;
    return (
        <html>
            {generateHeader(options)}
            <body>
                {navbar(options)}
                {progressbar(options)}
                {suiteList(options)}
            </body>
            {scripts(options)}
        </html>
    );
}
