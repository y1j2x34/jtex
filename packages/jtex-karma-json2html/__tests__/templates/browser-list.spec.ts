import { generateBrowserList } from '../../src/browser-list';
import { mockOptions } from '../common/mock-report-options';

it('render browser list correctly', () => {
    const options = mockOptions({});
    const browserListHTML = generateBrowserList(options);
    expect(browserListHTML).toMatchSnapshot();
});
