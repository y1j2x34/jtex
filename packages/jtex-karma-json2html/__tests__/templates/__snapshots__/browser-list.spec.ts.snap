// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render browser list correctly 1`] = `
"<div class=\\"dropdown jtex--dropdown\\"><button class=\\"btn btn-light dropdown-toggle\\" data-toggle=\\"dropdown\\" type=\\"button\\">Browsers</button>
<div class=\\"dropdown-menu switch-browser-dropdown-menu\\"><a class=\\"dropdown-item switch-browser-menu-item\\" href=\\"javascript:;\\" menu-browser=\\"82275798\\"><i class=\\"material-icons text-primary\\">check</i>
 
HeadlessChrome 87.0.4280 (Linux 0.0.0)</a>
<a class=\\"dropdown-item switch-browser-menu-item\\" href=\\"javascript:;\\" menu-browser=\\"82275790\\"><i class=\\"material-icons text-primary\\">check</i>
 
Chrome 87.0.4280 (Linux 0.0.0)</a></div></div>"
`;
